src/ui/training/utils.ts

```diff
@@ -0,0 +1,67 @@
+import * as xhr from './xhr'
+import { PuzzleSyncData, PuzzleData } from '../../lichess/interfaces/training'
+import router from '../../router'
+import settings from '../../settings'
+import { OfflinePuzzle } from './interfaces'
+
+export function syncPuzzles() {
+  const unsolved = settings.training.unsolvedPuzzles
+  const user = settings.training.user()
+  const curRating = user ? user.rating : null
+  const discardedPuzzles: OfflinePuzzle[] = []
+  // Cull puzzles that were retreived when the user's rating was significantly different than it is now
+  if (curRating)
+    unsolved(unsolved().reduce((acc: OfflinePuzzle[], cur: OfflinePuzzle) => {
+      const ratingDiff = (curRating && cur.userRating) ? Math.abs(curRating - cur.userRating) : 0
+      if (ratingDiff < settings.training.ratingDiffThreshold) {
+        discardedPuzzles.push(cur)
+        acc.push(cur)
+      }
+      return acc
+    }, []))
+
+  const puzzleDeficit = Math.max(settings.training.puzzleBufferLen - unsolved().length, 0)
+  let puzzlesLoaded = Promise.resolve(true)
+  if (puzzleDeficit) {
+    puzzlesLoaded = xhr.newPuzzles(puzzleDeficit).then((syncData: PuzzleSyncData) => {
+      unsolved(unsolved().concat(syncData.puzzles.map((puzzle: OfflinePuzzle) => {
+        if (syncData.user)
+          puzzle.userRating = syncData.user.rating
+        return puzzle
+      })))
+      settings.training.user(syncData.user)
+      return true
+    }).catch(() => {
+      // If loading new puzzles fails, add back the discarded ones to avoid potentially running out of puzzles
+      unsolved(unsolved().concat(discardedPuzzles))
+      return false
+    })
+  }
+  if (settings.training.solvedPuzzles().length) {
+    xhr.solvePuzzles(settings.training.solvedPuzzles()).then(() => settings.training.solvedPuzzles([]), () => {})
+  }
+  return puzzlesLoaded
+}
+
+function loadNextPuzzle() {
```

Comment:
it seems that this extra function is not necessary: you can put all the logic in `loadOfflinePuzzle()`